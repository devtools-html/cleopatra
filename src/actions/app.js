/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import type { ThreadIndex } from '../types/profile';
import type { CssPixels } from '../types/units';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export const reportTrackThreadHeight = (
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> => (dispatch, getState) => {
  const trackThreadHeights = getTrackThreadHeights(getState());
  const previousHeight = trackThreadHeights[threadIndex];
  if (previousHeight !== height) {
    // Guard against unnecessary dispatches. This could happen frequently.
    dispatch({
      type: 'UPDATE_TRACK_THREAD_HEIGHT',
      height,
      threadIndex,
    });
  }
};

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}
