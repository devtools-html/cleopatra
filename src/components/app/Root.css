.rootMessageContainer {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  width: 100%;
  height: 100%;
  align-items: center;
  justify-content: center;
}

.rootMessage {
  overflow: hidden;
  width: 80%;
  max-width: 600px;
  box-sizing: border-box;
  padding: 3em;
  border: 1px solid #ccc;
  background-color: #fff;
  border-radius: 3px;
  box-shadow: 0 5px 25px #0b1f50;
  font-size: 130%;
}

.rootMessageText,
.rootMessageAdditional {
  line-height: 1.5;
}

.rootMessageTitle {
  margin-top: 0;
}

.rootMessageAdditional {
  padding-top: 1em;
  margin-top: 1em;
  border-top: 1px solid #ccc;
  font-size: 12px;
}

.loading {
  position: relative;
  height: 40px;
  margin-top: 10px;
}

.loading-div {
  position: absolute;
  height: 8px;
  animation-duration: 4000ms;
  animation-iteration-count: infinite;
  animation-name: loadingdiv;
  border-radius: 2px;
}

.loading-row-1 {
  top: 0;
}

.loading-row-2 {
  top: 10px;
}

.loading-row-3 {
  top: 20px;
}

.loading-row-4 {
  top: 30px;
}

.loading-div-1 {
  width: 100%;
  animation-delay: -900ms;
  background-color: var(--red-50);
}

.loading-div-2 {
  left: 1%;
  width: 97%;
  animation-delay: -800ms;
  background-color: var(--blue-50);
}

.loading-div-3 {
  left: 5%;
  width: 30%;
  animation-delay: -700ms;
  background-color: var(--orange-50);
}

.loading-div-4 {
  left: 43%;
  width: 53%;
  animation-delay: -600ms;
  background-color: var(--purple-50);
}

.loading-div-5 {
  left: 8%;
  width: 14%;
  animation-delay: -500ms;
  background-color: var(--purple-50);
}

.loading-div-6 {
  left: 23%;
  width: 11%;
  animation-delay: -400ms;
  background-color: var(--green-50);
}

.loading-div-7 {
  left: 44%;
  width: 15%;
  animation-delay: -300ms;
  background-color: var(--orange-50);
}

.loading-div-8 {
  left: 60%;
  width: 17%;
  animation-delay: -200ms;
  background-color: var(--orange-50);
}

.loading-div-9 {
  left: 78%;
  width: 2%;
  animation-delay: -100ms;
  background-color: var(--blue-50);
}

.loading-div-10 {
  left: 81%;
  width: 11%;
  animation-delay: 0ms;
  background-color: var(--green-60);
}

@keyframes loadingdiv {
  0% {
    opacity: 0;
    transform: translateX(100px);
  }

  20% {
    opacity: 0;
    transform: translateX(100px);
  }

  45% {
    opacity: 1;
    transform: translateX(0);
  }

  65% {
    opacity: 1;
    transform: translateX(0);
  }

  100% {
    opacity: 0;
    transform: translateX(-100px);
  }
}
