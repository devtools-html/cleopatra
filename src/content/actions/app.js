// @flow
import type { Action } from './types';

export function changeSelectedTab(selectedTab: string): Action {
  return {
    type: 'CHANGE_SELECTED_TAB',
    selectedTab,
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}
