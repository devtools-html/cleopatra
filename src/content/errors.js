// @flow

type Attempt = {
  count: number,
  total: number,
};

export class TemporaryError extends Error {
  attempt: Attempt;

  constructor(message: string, attempt: Attempt) {
    super(message);
    this.name = 'TemporaryError';
    this.attempt = attempt;
  }
}
