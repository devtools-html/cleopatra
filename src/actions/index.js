/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
