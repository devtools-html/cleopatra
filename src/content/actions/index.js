// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
