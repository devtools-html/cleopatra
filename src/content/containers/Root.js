/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React, { PureComponent, PropTypes } from 'react';
import { connect, Provider } from 'react-redux';
import { oneLine } from 'common-tags';

import { retrieveProfileFromAddon, retrieveProfileFromWeb } from '../actions/receive-profile';
import ProfileViewer from '../components/ProfileViewer';
import Home from '../containers/Home';
import { urlFromState, stateFromLocation } from '../url-handling';
import { getView } from '../reducers/app';
import { getDataSource, getHash } from '../reducers/url-state';
import URLManager from './URLManager';

import type { Store } from '../types';
import type { AppViewState } from '../reducers/types';

const LOADING_MESSAGES = Object.freeze({
  'from-addon': 'Retrieving profile from the gecko profiler addon...',
  'from-file': 'Reading the file and parsing the profile in it...',
  'local': 'Not implemented yet.',
  'public': 'Retrieving profile from the public profile store...',
});

const ERROR_MESSAGES = Object.freeze({
  'from-addon': "Couldn't retrieve the profile from the gecko profiler addon.",
  'from-file': "Couldn't read the file or parse the profile in it.",
  'local': 'Not implemented yet.',
  'public': "Couldn't Retrieve the profile from the public profile store.",
});

// TODO Switch to a proper i18n library
function fewTimes(count: number) {
  switch (count) {
    case 1: return 'once';
    case 2: return 'twice';
    default: return `${count} times`;
  }
}

type ProfileViewProps = {
  view: AppViewState,
  dataSource: string,
  hash: string,
  retrieveProfileFromAddon: void => void,
  retrieveProfileFromWeb: string => void,
};

class ProfileViewWhenReadyImpl extends PureComponent {
  props: ProfileViewProps;

  componentDidMount() {
    const { dataSource, hash, retrieveProfileFromAddon, retrieveProfileFromWeb } = this.props;
    switch (dataSource) {
      case 'from-addon':
        retrieveProfileFromAddon();
        break;
      case 'from-file':
        // retrieveProfileFromFile should already have been called
        break;
      case 'local':
        break;
      case 'public':
        retrieveProfileFromWeb(hash);
        break;
    }
  }

  render() {
    const { view, dataSource } = this.props;
    switch (view.phase) {
      case 'INITIALIZING': {
        if (dataSource === 'none') {
          return <Home />;
        }

        const message = LOADING_MESSAGES[dataSource] || 'View not found';
        let additionalMessage = '';
        if (view.additionalData) {
          if (view.additionalData.message) {
            additionalMessage = view.additionalData.message;
          }

          if (view.additionalData.attempt) {
            const attempt = view.additionalData.attempt;
            additionalMessage += `Tried ${fewTimes(attempt.count)} out of ${attempt.total}.`;
          }
        }

        return (
          <div>
            <div>{ message }</div>
            { additionalMessage && <div>{ additionalMessage }</div>}
          </div>
        );
      }
      case 'FATAL_ERROR': {
        const message = ERROR_MESSAGES[dataSource] || "Couldn't retrieve the profile.";
        let additionalMessage = null;
        if (view.error) {
          console.error(view.error);
          additionalMessage = oneLine`
            Error was "${view.error.toString()}".
            The full stack has been written to the Web Console.
          `;
        }

        return (
          <div>
            <div>{ message }</div>
            { additionalMessage && <div>{ additionalMessage }</div>}
          </div>
        );
      }
      case 'PROFILE':
        return <ProfileViewer/>;
      case 'ROUTE_NOT_FOUND':
        return <div>There is no route handler for the URL {window.location.pathname + window.location.search}</div>;
      default:
        return <div>View not found.</div>;
    }
  }
}

ProfileViewWhenReadyImpl.propTypes = {
  view: PropTypes.shape({
    phase: PropTypes.string.isRequired,
    additionalData: PropTypes.object,
    error: PropTypes.instanceOf(Error),
  }).isRequired,
  dataSource: PropTypes.string.isRequired,
  hash: PropTypes.string,
  retrieveProfileFromAddon: PropTypes.func.isRequired,
  retrieveProfileFromWeb: PropTypes.func.isRequired,
};

const ProfileViewWhenReady = connect(
  state => ({
    view: getView(state),
    dataSource: getDataSource(state),
    hash: getHash(state),
  }),
  { retrieveProfileFromWeb, retrieveProfileFromAddon }
)(ProfileViewWhenReadyImpl);

type RootProps = {
  store: Store,
};

export default class Root extends PureComponent {
  props: RootProps;

  render() {
    const { store } = this.props;
    return (
      <Provider store={store}>
        <URLManager urlFromState={urlFromState} stateFromLocation={stateFromLocation}>
          <ProfileViewWhenReady/>
        </URLManager>
      </Provider>
    );
  }
}

Root.propTypes = {
  store: PropTypes.any.isRequired,
};
