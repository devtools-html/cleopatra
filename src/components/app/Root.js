/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React, { PureComponent } from 'react';
import { Provider } from 'react-redux';
import { UrlManager } from './UrlManager';
import { FooterLinks } from './FooterLinks';
import { ErrorBoundary } from './ErrorBoundary';
import { AppViewRouter } from './AppViewRouter';
import { ProfileLoader } from './ProfileLoader';
import { ServiceWorkerManager } from './ServiceWorkerManager';
import { WindowTitle } from './WindowTitle';
import { AppLocalizationProvider } from 'firefox-profiler/components/app/AppLocalizationProvider';
import { createBrowserConnection } from 'firefox-profiler/app-logic/browser-connection';

import type { BrowserConnectionStatus } from 'firefox-profiler/app-logic/browser-connection';
import type { Store } from 'firefox-profiler/types';

import './Root.css';

type RootProps = {
  store: Store,
};

type StateProps = {|
  browserConnectionStatus: BrowserConnectionStatus,
|};

import { DragAndDrop } from './DragAndDrop';

export class Root extends PureComponent<RootProps, StateProps> {
  state = {
    browserConnectionStatus: { status: 'WAITING' },
  };

  render() {
    const { store } = this.props;
    const { browserConnectionStatus } = this.state;

    const browserConnection =
      browserConnectionStatus.status === 'ESTABLISHED'
        ? browserConnectionStatus.browserConnection
        : null;

    return (
      <ErrorBoundary message="Uh oh, some error happened in profiler.firefox.com.">
        <Provider store={store}>
          <AppLocalizationProvider>
            <DragAndDrop browserConnection={browserConnection}>
              <UrlManager browserConnectionStatus={browserConnectionStatus}>
                <ServiceWorkerManager />
                <ProfileLoader
                  browserConnectionStatus={browserConnectionStatus}
                />
                <AppViewRouter
                  browserConnectionStatus={browserConnectionStatus}
                />
                <FooterLinks />
                <WindowTitle />
              </UrlManager>
            </DragAndDrop>
          </AppLocalizationProvider>
        </Provider>
      </ErrorBoundary>
    );
  }

  componentDidMount() {
    createBrowserConnection().then((browserConnectionStatus) => {
      this.setState({ browserConnectionStatus });
    });
  }
}
