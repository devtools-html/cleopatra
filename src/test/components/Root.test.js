/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// We want to test these components in isolation and tightly control the actions
// dispatched and avoid any side-effects.  That's why we mock this module and
// return dummy thunk actions that return a Promise.
jest.mock('../../actions/receive-profile', () => ({
  retrieveProfileFromAddon: () => async () => {},
  retrieveProfileFromStore: () => async () => {},
}));

import * as React from 'react';

import { ProfileViewWhenReady } from '../../components/app/Root';
import { updateUrlState } from '../../actions/app';
// Because this module is mocked but we want the real actions in the test, we
// use `jest.requireActual` here.
const {
  fatalError,
  temporaryError,
  viewProfile,
  waitingForProfileFromAddon,
  waitingForProfileFromStore,
} = jest.requireActual('../../actions/receive-profile');
import { stateFromLocation } from '../../app-logic/url-handling';
import { TemporaryError } from '../../utils/errors';

import { shallowWithStore } from '../fixtures/enzyme';
import { blankStore } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';

describe('app/ProfileViewWhenReady', function() {
  it('renders an initial home', function() {
    const { view } = setup();

    // dive() will shallow-render the wrapped component
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders the addon loading page', function() {
    const { view, dispatch, navigateToAddonLoadingPage } = setup();

    navigateToAddonLoadingPage();
    dispatch(waitingForProfileFromAddon());
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders the profile view', function() {
    const { view, dispatch, navigateToStoreLoadingPage } = setup();

    navigateToStoreLoadingPage();
    dispatch(waitingForProfileFromStore());
    expect(view.dive()).toMatchSnapshot();

    const { profile } = getProfileFromTextSamples(`A`);
    dispatch(viewProfile(profile));
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders temporary errors', function() {
    const { view, dispatch, navigateToStoreLoadingPage } = setup();

    navigateToStoreLoadingPage();
    dispatch(
      temporaryError(
        new TemporaryError('This is a temporary error, wait a bit more.', {
          count: 3,
          total: 10,
        })
      )
    );
    expect(view.dive()).toMatchSnapshot();
  });

  it('renders an home when the user pressed back after an error', function() {
    const {
      view,
      dispatch,
      navigateToStoreLoadingPage,
      navigateBackToHome,
    } = setup();

    navigateToStoreLoadingPage();
    dispatch(fatalError(new Error('Error while loading profile')));
    expect(view.dive()).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();

    navigateBackToHome();
    expect(view.dive()).toMatchSnapshot();
  });
});

function setup() {
  // Let's silence the error output to the console
  jest.spyOn(console, 'error').mockImplementation(() => {});

  const store = blankStore();
  const view = shallowWithStore(<ProfileViewWhenReady />, store);

  function navigateToStoreLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/public/ThisIsAFakeHash/calltree',
      search: '',
      hash: '',
    });
    store.dispatch(updateUrlState(newUrlState));
  }

  function navigateToAddonLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/from-addon/',
      search: '',
      hash: '',
    });
    store.dispatch(updateUrlState(newUrlState));
  }

  function navigateBackToHome() {
    const newUrlState = stateFromLocation({
      pathname: '/',
      hash: '',
      search: '',
    });
    store.dispatch(updateUrlState(newUrlState));
  }

  return {
    view,
    dispatch: store.dispatch,
    navigateToStoreLoadingPage,
    navigateToAddonLoadingPage,
    navigateBackToHome,
  };
}
