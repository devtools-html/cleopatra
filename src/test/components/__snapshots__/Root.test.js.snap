// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ProfileViewWhenReady renders an home when the user pressed back after an error 1`] = `
<div
  className="rootMessageContainer"
>
  <div
    className="rootMessage"
  >
    <h1
      className="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      className="rootMessageText"
    >
      Could not download the profile.
    </div>
    <div
      className="rootMessageAdditional"
    >
      <p
        key="0"
      >
        Error: Error while loading profile
      </p>
      <p
        key="1"
      >
        The full stack has been written to the Web Console.
      </p>
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders an home when the user pressed back after an error 2`] = `<Connect(Home) />`;

exports[`app/ProfileViewWhenReady renders an initial home 1`] = `<Connect(Home) />`;

exports[`app/ProfileViewWhenReady renders temporary errors 1`] = `
<div
  className="rootMessageContainer"
>
  <div
    className="rootMessage"
  >
    <h1
      className="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      className="rootMessageText"
    >
      Downloading and processing the profile...
    </div>
    <div
      className="rootMessageAdditional"
    >
      <p
        key="0"
      >
        This is a temporary error, wait a bit more.
      </p>
      <p
        key="1"
      >
        Tried 3 times out of 10.
      </p>
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
    <div
      className="loading"
    >
      <div
        className="loading-div loading-div-1 loading-row-1"
      />
      <div
        className="loading-div loading-div-2 loading-row-2"
      />
      <div
        className="loading-div loading-div-3 loading-row-3"
      />
      <div
        className="loading-div loading-div-4 loading-row-3"
      />
      <div
        className="loading-div loading-div-5 loading-row-4"
      />
      <div
        className="loading-div loading-div-6 loading-row-4"
      />
      <div
        className="loading-div loading-div-7 loading-row-4"
      />
      <div
        className="loading-div loading-div-8 loading-row-4"
      />
      <div
        className="loading-div loading-div-9 loading-row-4"
      />
      <div
        className="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders the addon loading page 1`] = `
<div
  className="rootMessageContainer"
>
  <div
    className="rootMessage"
  >
    <h1
      className="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      className="rootMessageText"
    >
      Grabbing the profile from the Gecko Profiler Addon...
    </div>
    <div
      className="loading"
    >
      <div
        className="loading-div loading-div-1 loading-row-1"
      />
      <div
        className="loading-div loading-div-2 loading-row-2"
      />
      <div
        className="loading-div loading-div-3 loading-row-3"
      />
      <div
        className="loading-div loading-div-4 loading-row-3"
      />
      <div
        className="loading-div loading-div-5 loading-row-4"
      />
      <div
        className="loading-div loading-div-6 loading-row-4"
      />
      <div
        className="loading-div loading-div-7 loading-row-4"
      />
      <div
        className="loading-div loading-div-8 loading-row-4"
      />
      <div
        className="loading-div loading-div-9 loading-row-4"
      />
      <div
        className="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders the profile view 1`] = `
<div
  className="rootMessageContainer"
>
  <div
    className="rootMessage"
  >
    <h1
      className="rootMessageTitle"
    >
      perf.html
    </h1>
    <div
      className="rootMessageText"
    >
      Downloading and processing the profile...
    </div>
    <div
      className="loading"
    >
      <div
        className="loading-div loading-div-1 loading-row-1"
      />
      <div
        className="loading-div loading-div-2 loading-row-2"
      />
      <div
        className="loading-div loading-div-3 loading-row-3"
      />
      <div
        className="loading-div loading-div-4 loading-row-3"
      />
      <div
        className="loading-div loading-div-5 loading-row-4"
      />
      <div
        className="loading-div loading-div-6 loading-row-4"
      />
      <div
        className="loading-div loading-div-7 loading-row-4"
      />
      <div
        className="loading-div loading-div-8 loading-row-4"
      />
      <div
        className="loading-div loading-div-9 loading-row-4"
      />
      <div
        className="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/ProfileViewWhenReady renders the profile view 2`] = `<Connect(ProfileViewer) />`;
